### ISSUE TITLE
Mypy thinks __doc__ may be None even for classes that have a docstring

### ISSUE THREAD

**Issue by mthuurne** [python/mypy#9170]
When mypy 0.782 checks the following code:
```py
class C:
    """I'm documented."""

    def __str__(self) -> str:
        return self.__doc__
```
It reports:
```
error: Incompatible return value type (got "Optional[str]", expected "str")  [return-value]
```
I expected no error to be reported.

In general, `Optional[str]` is a good type for `__doc__` in Python 3, but if the type of the object on which `__doc__` is accessed is known, it can be possible to produce a more narrow type: `str` in the example above.

This issue might be related to #2709.

**Comment by Akuli** [comment]
`__doc__` really can be None. Try adding `print(C.__doc__)` to the end of your example and running `python3 -OO file.py`. However, there are other things that can be None but mypy doesn't complain about them.

For example, `sys.stdin`, `sys.stdout` and `sys.stderr` are None when running with `pythonw.exe` on Windows. `print("hello")` won't fail because print knows how to handle that, but `sys.stdout.write("hello\n")` would fail. Still, requiring all code using `sys.stdout.write` to `assert sys.stdout is not None` would be annoying.

I'm not a mypy developer, but I won't be surprised if this is a feature rather than a bug. Maybe supporting the `-OO` option is considered to be important.

To make your code type check correctly when Python's `-OO` switch isn't used, add

```
assert self.__doc__ is not None
```

before returning `self.__doc__`.

**Comment by mthuurne** [comment]
I didn't realize that `-OO` removes docstrings.

Since `-OO` also removes asserts, the example would return `None` when running with that flag, which could cause problems. So I'll have to find a different solution, but that shouldn't be too hard.

You make a good point about how it can be practical to ignore `None` in some cases, but a program introspecting its docstrings is probably not common enough to make such an exception for. Especially since most introspections will happen on types that are not statically known and therefore `None` cannot be ruled out even if `-OO` is ignored.
